Android only upgrade react-native to 0.72.x

siddarthkay opened the pull request:
This PR is just made to test and fix effects of this upgrade on CI.
Note: This PR only contains Android side of the upgrades.
state: WIP

siddarthkay committed:
(0.72.3) : semi aggressively nuke old code

siddarthkay committed:
upgrade react-native to 0.70.13

siddarthkay committed:
chore: drop iOS & e2e Jenkinsfiles

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :x: | afc5a939 | [#1](https://ci.status.im/job/status-mobile/job/prs/job/ios/job/PR-17062/1/) | 2023-08-19 06:54:39 | ~4 min | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-mobile/job/prs/job/ios/job/PR-17062/1/consoleText) |

siddarthkay committed:
upgrade `react-native-keychain` to 8.1.2

siddarthkay committed:
patch react-native-keychain library

jakubgs commented:
This failure:
```
A problem occurred configuring project ':react-native-keychain'.
> Could not determine the dependencies of null.
   > Could not resolve all task dependencies for configuration ':react-native-keychain:classpath'.
      > Could not resolve com.adarshr:gradle-test-logger-plugin:2.0.0.
        Required by:
            project :react-native-keychain
         > No cached version of com.adarshr:gradle-test-logger-plugin:2.0.0 available for offline mode
```
Appears to indicate lack of `com.adarshr:gradle-test-logger-plugin:2.0.0`, but it is there:
```
 > grep -A3 gradle-test-logger-plugin nix/deps/gradle/deps.json
    "path": "com/adarshr/gradle-test-logger-plugin/2.0.0",
    "repo": "https://plugins.gradle.org/m2",
    "files": {
      "gradle-test-logger-plugin-2.0.0.pom": {
        "sha1": "fe5c29ddcae80215fd7ec74476880de67d66c894",
        "sha256": "sha256-kVCmaVqB+SN+4958dClo5XcMHyWCGsS3c5kCFh9lUu4="
      },
      "gradle-test-logger-plugin-2.0.0.jar": {
        "sha1": "957ca7f6f4390ea8cf4551fff43a96da50844b98",
        "sha256": "sha256-lZFMdk+bW0P59xiCWMxr9BsLr2dnlhqjQIJtgHyAjp0="
      }
```
And it is in the store:
```
 > find /nix/store -maxdepth 1 -iname '*gradle-test-logger-plugin*'
/nix/store/c1l61jfc12wyan718s0hqgbanm7bq1pw-gradle-test-logger-plugin-2.0.0.pom.drv
/nix/store/cmsnv0x1jlk0i52w8rnbxxg60891vn7i-gradle-test-logger-plugin-2.0.0.jar.drv
/nix/store/rdyp3w7s5qpzi9afbzi53gigxk2kr417-gradle-test-logger-plugin-2.0.0.jar
/nix/store/n8gsbfnncklycmx689ibsmplkf7rs9gq-gradle-test-logger-plugin-2.0.0.pom
```
But if I `strace` Gradle at build time(and use `--keep-failed` flag for `nix-build`) I see no attempt to find it on the filesystem:
```
/tmp/nix-build-status-mobile-build-pr-android.drv-0
 > grep gradle-test-logger-plugin android/strace.log | wc -l
0
```
Which together with the `null` in `Could not determine the dependencies of null.` suggests some deeper issues.

jakubgs commented:
Also, this change(https://github.com/status-im/status-mobile/pull/17062/commits/67a4164462833051f5bb461e490a297bce85ab6f) appears to be entirely unnecessary:
https://github.com/status-im/status-mobile/blob/67a4164462833051f5bb461e490a297bce85ab6f/nix/deps/gradle/generate.sh#L89
Because the dependency is included even without it:
```
~/work/status-mobile upgrade/android-only-rn-0.72.x ⇣
 > g
## upgrade/android-only-rn-0.72.x...origin/upgrade/android-only-rn-0.72.x [behind 1]

~/work/status-mobile upgrade/android-only-rn-0.72.x ⇣
 > grep gradle-test-logger-plugin nix/deps/gradle/deps.json
    "path": "com/adarshr/gradle-test-logger-plugin/2.0.0",
      "gradle-test-logger-plugin-2.0.0.pom": {
      "gradle-test-logger-plugin-2.0.0.jar": {
```
As far as I can tell the key issue here is the `null` in `Could not determine the dependencies of null.`, because if you look at other examples of such errors the name of the package should be there:

* https://github.com/status-im/status-mobile/issues/13320
* https://github.com/status-im/status-mobile/pull/13337

siddarthkay commented:
@jakubgs maybe this is related : https://github.com/oblador/react-native-keychain/issues/595#issuecomment-1568657311

jakubgs commented:
Yes, it's most likely that the issue is not in `react-native-keychain` but in `gradle-test-logger-plugin`, specifically `2.0.0`.

siddarthkay committed:
upgrade react-native to 0.72.2

yakimant commented on android/build.gradle:
```diff
@@ -66,32 +55,26 @@ allprojects {
+//         exclusiveContent {
+//                    filter {
+//                        includeGroup "com.facebook.react"
+//                    }
+//                    forRepository {
+//                        maven {
+//                            url "$rootDir/../node_modules/react-native/reactAndroid"
+//                        }
+//                    }
+//         }
+//         mavenLocal()
         // Android JSC is installed from npm
         maven { url("$rootDir/../node_modules/jsc-android/dist") }
         // For geth, function, and status-go
         flatDir { dirs "libs", System.env.STATUS_GO_ANDROID_LIBDIR }
         // Everything Else
-        google()
-        jcenter() // Required for @react-native-community/blur:3.6.1, can be removed after updating to 3.6.1+
-        mavenCentral()
+//         google()
+//         jcenter() // Required for @react-native-community/blur:3.6.1, can be removed after updating to 3.6.1+
+//         mavenCentral()
         //         maven { url "https://www.jitpack.io" }
-        maven { url 'https://mvnrepository.com/artifact/com.github.gundy/semver4j'}
+//         maven { url 'https://mvnrepository.com/artifact/com.github.gundy/semver4j'}
     }
 }
```
This doesn't look right to me. What was the reason of removing?
Fixing some build issues?

yakimant replied:
This is for debug purpose to try to fix an issue, which id on CI/build-android currently.

siddarthkay committed:
chore: add missing deps

siddarthkay committed:
debug: add missing deps

yakimant commented:
`url2json.sh` has an issue with new package.

https://repo1.maven.org/maven2/com/facebook/react/react-android/0.72.3/react-android-0.72.3.pom
has type `pom`, but it actually contains `aar` and `jar` files:
https://repo1.maven.org/maven2/com/facebook/react/react-android/0.72.3/

This code need to be ammended:
https://github.com/status-im/status-mobile/blob/255a3b917265d381fed57e26422277b4cc8772e4/nix/deps/gradle/url2json.sh#L117-L123

yakimant committed:
fix react-android and attempt to fix prefab cli issue

yakimant commented:
The latest issue with gradel deps:
```
Execution failed for task ':react-native-reanimated:configureCMakeRelWithDebInfo[arm64-v8a]'.
> Could not resolve all files for configuration ':react-native-reanimated:_internal_prefab_binary'.
   > Could not find cli-2.0.0-all.jar (com.google.prefab:cli:2.0.0).
     Searched in the following locations:
         file:/nix/store/2vlvms16hq90rz31qb9zf8i2dhdvjcm0-status-mobile-maven-deps/com/google/prefab/cli/2.0.0/cli-2.0.0-all.jar
```

siddarthkay committed:
debug: fix reanimated 3 build issues

siddarthkay committed:
debug: fix async storage lib deps

siddarthkay committed:
upgrade react-native-async-storage to 1.19.3

siddarthkay committed:
bump aapt2 to `7.4.2-8841542`

siddarthkay committed:
fix: make run-android needs debug.aar for RN

jakubgs commented:
Lol, what the fuck is this?

![image](https://github.com/status-im/status-mobile/assets/2212681/dfe23925-a433-4d13-9aa0-4591eb318075)

Why does it have a a `-release` suffix? that's not something I've ever seen. Are you sure Gradle actually looks for this?

yakimant commented on android/gradle.properties:
```diff
@@ -42,7 +42,7 @@ KEYSTORE_KEY_PASSWORD=password
 # By default we build a mostly universal APK
 ANDROID_ABI_SPLIT=false
 # Some platforms are excluded though
-ANDROID_ABI_INCLUDE=armeabi-v7a;arm64-v8a;x86
```
This can be overriten by Make/Jenkins & Nix in some placces.

jakubgs replied:
That's fine. It's still a sensible default necessary to be able to use Gradle without all the baggage.

jakubgs replied:
Although I do wonder why this architecture is being removed.

siddarthkay replied:
Initially I removed it to get builds to work because of this : https://github.com/software-mansion/react-native-reanimated/issues/4715#issuecomment-1636855035
Will validate if its needed anymore or not during PR cleanup!

jakubgs commented on nix/deps/gradle/url2json.sh:
```diff
@@ -121,5 +121,9 @@ echo -ne "
+[[ "${PKG_NAME}" == react-android-* ]] && fetch_and_template_file "${PKG_NAME}.module"
+[[ "${PKG_NAME}" == react-android-* ]] && fetch_and_template_file "${PKG_NAME}-release.aar"
+[[ "${PKG_NAME}" == react-android-* ]] && fetch_and_template_file "${PKG_NAME}-debug.aar"
+[[ "${PKG_NAME}" == cli-* ]] && fetch_and_template_file "${PKG_NAME}-all.jar"
```
That's horrible. I don't agree with this change at all. This needs to be dealt with in a generic way. Adding more special-case exceptions is just bad.

siddarthkay replied:
this okay? : https://github.com/status-im/status-mobile/pull/17062/commits/ef67791ba3857b0ceba80ed0fd3d46f1dea5985c

yakimant replied:
I added it to test that it helped.
As for packaging style changed, note:
- packaging is pom, but it has aars
- it's not only pom and aar, but also module
I wonder if it's react-native specific, or another way of handling maven packages. Then we could make it even more generic for all new cases, which might appear in the future.

jakubgs commented on nix/deps/nodejs-patched/default.nix:
```diff
@@ -89,6 +69,26 @@ stdenv.mkDerivation {
+#  TODO: remove this patch when we this library fixes above issue
+  patchKeyChainLibraryPhase = ''
+   sed -i -e '/classpath/d' \
+          -e '/apply plugin: "com\.adarshr\.test-logger"/d' ./node_modules/react-native-keychain/android/build.gradle
```
```suggestion
          -e '/apply plugin: "com\.adarshr\.test-logger"/d' \
          ./node_modules/react-native-keychain/android/build.gradle
```

jakubgs commented on nix/deps/nodejs-patched/default.nix:
```diff
@@ -89,6 +69,26 @@ stdenv.mkDerivation {
 
+#  Remove gradle-test-logger-plugin:
+#  https://github.com/oblador/react-native-keychain/issues/595
+#  TODO: remove this patch when we this library fixes above issue
```
Why is there no indent?

jakubgs commented on nix/deps/nodejs-patched/default.nix:
```diff
@@ -89,6 +69,26 @@ stdenv.mkDerivation {
+#  Error : CKCamera.kt: (183, 17): 'onScale' overrides nothing
+#  fix from : https://github.com/teslamotors/react-native-camera-kit/issues/535#issuecomment-1506229244
+#  note by library author : https://github.com/teslamotors/react-native-camera-kit/pull/551#issuecomment-1615305719
+#  TODO: remove this patch when we react-native-camera-kit releases a stable v14
```
Why is there no indent?

jakubgs commented on nix/deps/nodejs-patched/default.nix:
```diff
@@ -9,10 +9,9 @@ stdenv.mkDerivation {
-    "patchHermesPhase"
-    "patchJavaPhase"
     "patchReactNativePhase"
+    "patchKeyChainLibraryPhase"
```
Do we need the `Library` word in this? Seems to add nothing.

jakubgs commented on nix/deps/gradle/gradle_parser.awk:
```diff
@@ -26,7 +26,13 @@ function findPackage(line, regex) {
-        if (line ~ "com.facebook.react:(react-native|hermes-engine)") { continue }
+        # if (line ~ "com.facebook.react:(react-native|hermes-engine)") { continue }
+        # Example: +--- com.facebook.react:react-native:+ -> com.facebook.react:react-android:0.72.3
+        if (findPackage(line, "--- [^ ]+ -> ([^ :]+):([^ :]+):([^ :]+)$")) {
```
Eeeeee... Isn't this already handled by the last `findPackage` call?
https://github.com/status-im/status-mobile/blob/1203834b8c0dacf8931bb71d08a4777ec9a67415/nix/deps/gradle/gradle_parser.awk#L47-L48
What's different about this?

siddarthkay replied:
I am re-verifying this, will update here.

jakubgs commented on nix/deps/gradle/gradle_parser.awk:
```diff
@@ -26,7 +26,13 @@ function findPackage(line, regex) {
+            continue
+        }
+
+        if (line ~ "org.webkit:android-jsc") { continue }
```
Make this first, and add a comment why it is necessary.

jakubgs commented on nix/deps/gradle/generate.sh:
```diff
@@ -83,9 +83,15 @@ fi
-echo 'org.jetbrains.kotlin.jvm:org.jetbrains.kotlin.jvm.gradle.plugin:1.6.10' >> "${DEPS_LIST}"
-echo 'com.android.tools.lint:lint-gradle:30.2.2' >> "${DEPS_LIST}"
-echo 'com.android.tools.build:gradle:7.0.4' >> "${DEPS_LIST}"
+echo 'com.android.tools.build:gradle:1.3.1' >> "${DEPS_LIST}"
+echo 'com.squareup.okio:okio:1.13.0' >> "${DEPS_LIST}"
+echo 'com.squareup.okio:okio:1.15.0' >> "${DEPS_LIST}"
+echo 'com.squareup.okhttp3:okhttp:3.12.1' >> "${DEPS_LIST}"
+echo 'org.jetbrains.kotlin.jvm:org.jetbrains.kotlin.jvm.gradle.plugin:1.7.22' >> "${DEPS_LIST}"
+echo 'com.android.tools.lint:lint-gradle:30.4.2' >> "${DEPS_LIST}"
+echo 'com.android.tools.build:gradle:7.4.2' >> "${DEPS_LIST}"
+# Required by configuration ':react-native-async-storage_async-storage:detachedConfiguration2'.
+echo 'com.android.tools.build:aapt2:7.4.2-8841542' >> "${DEPS_LIST}"
```
Merciful Christ. This is awful. Are we sure we need all of those? We seriously need to deal with this issue:

* https://github.com/status-im/status-mobile/issues/15447

yakimant replied:
I agree. Do we have at lease a mechanism to alert if version is not in sync with required one?
Not to say, that we will still download a lib even if we stop using it, but I guess it's minor.

jakubgs commented on android/build.gradle:
```diff
@@ -64,34 +53,24 @@ allprojects {
             throw new GradleException('STATUS_GO_ANDROID_LIBDIR environment variable is not valid!')
         }
     }
+// Attempt to fix : error: package com.facebook.react.bridge does not exist while building react-native app
```
Probably no longer necessary right?

jakubgs commented on android/build.gradle:
```diff
@@ -35,11 +27,8 @@ buildscript {
     dependencies {
         classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:${project.ext.kotlinPluginVersion}"
-        classpath "com.android.tools.build:gradle:${project.ext.gradlePluginVersion}"
+        classpath("com.android.tools.build:gradle")
```
Why are we skipping the version?

siddarthkay replied:
we no longer need to specify version, ref : https://github.com/facebook/react-native/blob/0.72-stable/packages/react-native/template/android/build.gradle#L18

jakubgs commented on android/app/build.gradle:
```diff
@@ -1,97 +1,25 @@
  */
 
-def disableHermes = System.getenv('DISABLE_HERMES') == 'true'
+// def disableHermes = System.getenv('DISABLE_HERMES') == 'true'
```
Remove it?

jakubgs commented on android/app/build.gradle:
```diff
@@ -134,15 +110,8 @@ def jscFlavor = 'org.webkit:android-jsc:+'
  * and the benefits of using Hermes will therefore be sharply reduced.
  */
-def enableHermes = project.ext.react.get("enableHermes", false);
+def enableHermes = hermesEnabled.toBoolean();
```
What does this do?

siddarthkay replied:
sets it to whatever was set in `gradle.properties`, in our case `false`
according to new template : https://github.com/facebook/react-native/blob/0.72-stable/packages/react-native/template/android/gradle.properties#L44

jakubgs commented on android/app/build.gradle:
```diff
@@ -384,54 +277,32 @@ android {
+    // The version of react-native is set by the React Native Gradle Plugin
+    implementation("com.facebook.react:react-android")
+//     implementation 'com.facebook.soloader:soloader:0.10.4+'
+//     implementation("androidx.swiperefreshlayout:swiperefreshlayout:1.0.0")
```
If we don't need it lets remove it.

jakubgs reviewed (commented):
Good work man. But we need to do something about all those gradle exceptions added.

siddarthkay commented:
@jakubgs : Thanks for the review! I will clean up this PR 👍🏻

siddarthkay committed:
fixed `node-libs-react-native/globals` import

siddarthkay committed:
debug: build for `armeabi-v7a` arch as well

siddarthkay committed:
chore: refactor `url2json.sh` to be generic :)

siddarthkay committed:
chore: fix indents in patches

jakubgs commented on nix/deps/gradle/url2json.sh:
```diff
@@ -14,6 +14,12 @@ declare -a REPOS=(
   "https://jitpack.io"
 )
 
+# Special rule set for changes introduced by react-native 0.72
+# mapping package name patterns to file types
+declare -A RULE_SET
+RULE_SET["react-android-*"]="*.module *-release.aar *-debug.aar"
+RULE_SET["cli-*"]="*-all.jar"
```
I don't think this is a good solution. I think the correct solution is to check if `.module` exists, based on this:

>The .module file is a [Gradle Module Metadata](https://docs.gradle.org/current/userguide/publishing_gradle_module_metadata.html) file. It allows to publish two variants for debug and release.

https://github.com/react-native-community/discussions-and-proposals/blob/4a06fc6497c8e88d01502b0721d1c814ebc1fb61/proposals/0508-out-of-npm-artifacts.md#the-react-native-android-archive

From:

* https://github.com/react-native-community/discussions-and-proposals/pull/508

This shows us that if next to a [`react-android-0.72.3.pom`](https://repo1.maven.org/maven2/com/facebook/react/react-android/0.72.3/react-android-0.72.3.pom) exists a [`react-android-0.72.3.module`](https://repo1.maven.org/maven2/com/facebook/react/react-android/0.72.3/react-android-0.72.3.module) that would mean that the `-release` and `-debug`.

But honestly, I'm starting to think it might just be simpler to get a folder listing and fetch most files without any logic attached to it, since it just makes things more complex than they need to be.

jakubgs replied:
One problem though is that `dl.google.com` repo does not provide listings of files:
```
 > curl -sv https://dl.google.com/dl/android/maven2/com/google/prefab/cli/2.0.0/ 2>&1 > /dev/null  | grep '^<'
< HTTP/2 404 
< content-length: 1449
< content-type: text/html; charset=utf-8
< server: downloads
```

jakubgs replied:
I could't find a way to list things from `dl.google.com`. If you know or can find a way to do that we could then simplify `url2json.sh` script a lot by simply downloading everything listed(or almost everything).

yakimant replied:
https://maven.google.com/web/index.html#com.google.prefab:cli:2.0.0
as mentioned by @siddarthkay, has the links to the artifacts.

jakubgs replied:
A separate system for listing things. That's fucking great. Instead of simplifying it they are complicating it.

jakubgs replied:
@siddarthkay in hindsight I think your original solution was better. The `handle_react_native_new_dependencies` doesn't make things simpler, it makes them more obscure.

jakubgs replied:
I found something nice by examining the requests in the dev console:
```
 > curl -s https://dl.google.com/android/maven2/com/google/prefab/cli/2.0.0/artifact-metadata.json  | jq
{
  "artifacts": [
    {
      "name": "cli-2.0.0-all.jar",
      "tag": "jar"
    },
    {
      "name": "cli-2.0.0.pom",
      "tag": "pom"
    }
  ]
}
```
I think there's hope.

siddarthkay committed:
pin to android-jsc found in mavenCentral

\

siddarthkay committed:
upgrade react-native to 0.72.4

siddarthkay committed:
debug: get the app working in debug mode

siddarthkay committed:
fix: re-animated v3 animations cleanup

- Easing Import
- FadeIn swap for Transition Import
- easing
- linear
- Extrapolate
- Interpolate

siddarthkay committed:
fix: node deps mess during rebase :)

flexsurfer commented:
looks stale, feel free to reopen if needed
